torch/testing/asserts.py

```diff
@@ -0,0 +1,186 @@
+import functools
+import inspect
+import warnings
+from typing import Callable, Optional, Tuple
+
+import torch
+from torch.testing._internal.common_utils import get_comparison_dtype as _get_comparison_dtype, TestCase as _TestCase
+
+__all__ = ["assert_tensors_equal", "assert_tensors_allclose"]
+
+try:
+    import pytest
+
+    UsageError = pytest.UsageError
+except ImportError:
+
+    class UsageError(Exception):
+        pass
+
+
+class UsageWarning(Warning):
+    pass
+
+
+def _get_default_rtol_and_atol(tensor1: torch.Tensor, tensor2: torch.Tensor) -> Tuple[float, float]:
+    return _TestCase()._getDefaultRtolAndAtol(tensor1.dtype, tensor2.dtype)
+
+
+def _hide_internal_traceback(fn):
```

Comment:
This is a decorator that we can put onto every user facing function that automatically hides the internal traceback of our functions when using `pytest`. The user only sees the error with message as if it happened directly in the test.

Reply:
Add a comment here explaining that

Reply:
Would you post a comment showing the before/after of pytest output without/with this feature?

Reply:
Running

```python
def foo():
    assert False, "Failed because of reasons"


def bar():
    foo()


def baz():
    bar()


def test_baz():
    baz()


def spam():
    __tracebackhide__ = True
    assert False, "Failed because of reasons"


def ham():
    __tracebackhide__ = True
    spam()


def eggs():
    ham()


def test_eggs():
    eggs()
```

with `pytest --assert=plain` (otherwise `pytest` will add information to the `assert`'s since they are defined in the test module rather in an external one) outputs:

```
________________________________________ test_baz _________________________________________

    def test_baz():
>       baz()

test_traceback.py:14: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test_traceback.py:10: in baz
    bar()
test_traceback.py:6: in bar
    foo()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def foo():
>       assert False, "Failed because of reasons"
E       AssertionError: Failed because of reasons

test_traceback.py:2: AssertionError
________________________________________ test_eggs ________________________________________

    def test_eggs():
>       eggs()

test_traceback.py:32: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def eggs():
>       ham()
E       AssertionError: Failed because of reasons

test_traceback.py:28: AssertionError
```

As you can see the (internal helper functions) `foo` and `bar` as part of the traceback, whereas `ham` and `eggs` are hidden. If our final error message is expressive, IMO we don't need to hit the user with a mile long traceback that shows our internal workings.

The decorator works by assigning `__tracebackhide__ = True` into each previous frame in case it encounters an `AssertionError`. This is a [`pytest` feature](https://docs.pytest.org/en/stable/example/simple.html?highlight=__tracebackhide__#writing-well-integrated-assertion-helpers), so we might want to check if the test is run by `pytest` before we do that.

Reply:
Checking for pytest sounds good.

Reply:
I think this should be done as a manipulation in conftest.py since that is pytest-specific, without adding any special handling to this file.

Reply:
This only works for us. Given that we want to enable libraries that live on top of pytorch to use these, they also need to replicate our `conftest.py`.